### ISSUE TITLE
Continues respawn of getty tty

### ISSUE THREAD

**Issue by derkd** [docker-library/openjdk#107]
Hi all, 

I get a continues stream of log entries like below. 

```shell
Mar 26 04:56:36 node27091-nnc-prod daemon.info init: can't open /dev/tty5: No such file or directory
Mar 26 04:56:36 node27091-nnc-prod daemon.info init: can't open /dev/tty6: No such file or directory
Mar 26 04:56:37 node27091-nnc-prod daemon.info init: process '/sbin/getty 38400 tty1' (pid 2750) exited. Scheduling for restart.
Mar 26 04:56:37 node27091-nnc-prod daemon.info init: process '/sbin/getty 38400 tty2' (pid 2751) exited. Scheduling for restart.
Mar 26 04:56:37 node27091-nnc-prod daemon.info init: process '/sbin/getty 38400 tty3' (pid 2752) exited. Scheduling for restart.
Mar 26 04:56:37 node27091-nnc-prod daemon.info init: process '/sbin/getty 38400 tty4' (pid 2753) exited. Scheduling for restart.
Mar 26 04:56:37 node27091-nnc-prod daemon.info init: process '/sbin/getty 38400 tty5' (pid 2754) exited. Scheduling for restart.
Mar 26 04:56:37 node27091-nnc-prod daemon.info init: process '/sbin/getty 38400 tty6' (pid 2755) exited. Scheduling for restart.
Mar 26 04:56:37 node27091-nnc-prod daemon.info init: can't open /dev/tty1: No such file or directory
Mar 26 04:56:37 node27091-nnc-prod daemon.info init: starting pid 2757, tty '/dev/tty2': '/sbin/getty 38400 tty2'
Mar 26 04:56:37 node27091-nnc-prod daemon.info init: can't open /dev/tty3: No such file or directory
Mar 26 04:56:37 node27091-nnc-prod daemon.info init: can't open /dev/tty4: No such file or directory
Mar 26 04:56:37 node27091-nnc-prod daemon.info init: can't open /dev/tty5: No such file or directory
Mar 26 04:56:37 node27091-nnc-prod daemon.info init: can't open /dev/tty6: No such file or directory
```

I fixed this by adding this line in my dockerfile

```
#comment all tty stuff in /etc/inittab
RUN sed -i -- 's/tty/#tty/g' /etc/inittab
```

This now only resulted in the log below but at least it's only once and not every second

```shell
Apr 10 19:32:29 localhost daemon.info init: can't open /dev/tty5: No such file or directory
Apr 10 19:32:29 localhost daemon.info init: can't open /dev/tty6: No such file or directory
Apr 10 19:32:30 localhost daemon.info init: process '/sbin/getty 38400 tty1' (pid 280) exited. Scheduling for restart.
Apr 10 19:32:30 localhost daemon.info init: process '/sbin/getty 38400 tty2' (pid 281) exited. Scheduling for restart.
Apr 10 19:32:30 localhost daemon.info init: process '/sbin/getty 38400 tty3' (pid 282) exited. Scheduling for restart.
Apr 10 19:32:30 localhost daemon.info init: process '/sbin/getty 38400 tty4' (pid 283) exited. Scheduling for restart.
Apr 10 19:32:30 localhost daemon.info init: process '/sbin/getty 38400 tty5' (pid 284) exited. Scheduling for restart.
Apr 10 19:32:30 localhost daemon.info init: process '/sbin/getty 38400 tty6' (pid 285) exited. Scheduling for restart.
Apr 10 19:32:30 localhost daemon.info init: can't open /dev/tty1: No such file or directory
Apr 10 19:32:30 localhost daemon.info init: can't open /dev/tty2: No such file or directory
Apr 10 19:32:30 localhost daemon.info init: can't open /dev/tty3: No such file or directory
Apr 10 19:32:30 localhost daemon.info init: can't open /dev/tty4: No such file or directory
Apr 10 19:32:30 localhost daemon.info init: can't open /dev/tty5: No such file or directory
Apr 10 19:32:30 localhost daemon.info init: can't open /dev/tty6: No such file or directory
Apr 10 19:32:31 localhost daemon.info init: starting pid 300, tty '': '/sbin/rc shutdown'
Apr 10 19:32:31 localhost syslog.info syslogd exiting
Apr 10 21:33:25 node26473-nnc-dev syslog.info syslogd started: BusyBox v1.25.1
Apr 10 19:33:25 node26473-nnc-dev auth.info sshd[391]: Server listening on 0.0.0.0 port 22.
Apr 10 19:33:25 node26473-nnc-dev auth.info sshd[391]: Server listening on :: port 22.
Apr 10 19:33:43 node26473-nnc-dev auth.info sshd[391]: Received signal 15; terminating.
Apr 10 19:33:43 node26473-nnc-dev auth.info sshd[563]: Server listening on 0.0.0.0 port 22.
Apr 10 19:33:43 node26473-nnc-dev auth.info sshd[563]: Server listening on :: port 22.
Apr 10 19:33:45 node26473-nnc-dev auth.err sshd[612]: error: Bind to port 22 on 0.0.0.0 failed: Address in use.
Apr 10 19:33:45 node26473-nnc-dev auth.err sshd[612]: error: Bind to port 22 on :: failed: Address in use.
Apr 10 19:33:45 node26473-nnc-dev auth.crit sshd[612]: fatal: Cannot bind any address.
Apr 10 19:33:56 node26473-nnc-dev auth.err sshd[954]: error: Bind to port 22 on 0.0.0.0 failed: Address in use.
Apr 10 19:33:56 node26473-nnc-dev auth.err sshd[954]: error: Bind to port 22 on :: failed: Address in use.
Apr 10 19:33:56 node26473-nnc-dev auth.crit sshd[954]: fatal: Cannot bind any address.
```

Can we further optimise this?

**Comment by yosifkit** [comment]
> # 1: Don't boot init
>
> Containers model processes not machines. Even if you think that you need to do this you are probably wrong. 
> \- http://crosbymichael.com/dockerfile-best-practices-take-2.html

Containers aren't really meant for running regular `init` in them.  Init does much more than is necessary for running a process in a container.  If you are trying to run multiple processes in one container, you should either look to split it into separate containers and use service discovery, or if that is impossible then something like http://supervisord.org/ should work for starting and monitoring multiple processes within one container.

**Comment by derkd** [comment]
Thanks!!!
How can I disable boot init?

**Comment by yosifkit** [comment]
How are you running your container now? What does you Dockerfile look like?  If you have a java source for your image, then it should be something like the following.

```console
$ # not entirely sure on what normal paths would be
$ cat Dockerfile
FROM openjdk:8-jdk
COPY . /usr/src/myapp
WORKDIR /usr/src/myapp
RUN javac Main.java
CMD ["java", "Main"]
$ docker build -t my-rest-image .
$ docker run -d --name java-rest-container my-rest-image
```

Can probably drop the `javac` if you just have some jar/war files I think.  This container will only have the JVM with the specified Class running.

**Comment by derkd** [comment]
I run them in Jelastic cloud which supports docker containers. 

My docker file looks a bit like this

FROM openjdk:8u121-jre-alpine

ENV JELASTIC_EXPOSE FALSE
ENV TIMEZONE Europe/Amsterdam
RUN apk add --update tzdata \
	cp /usr/share/zoneinfo/${TIMEZONE} /etc/localtime \
	echo "${TIMEZONE}" &gt; /etc/timezone  \
	apk del tzdata

COPY web.jar /app/web.jar

#Adding fonts
RUN apk add ttf-dejavu

#comment all tty stuff in /etc/inittab
RUN sed -i -- 's/tty/#tty/g' /etc/inittab

CMD java -jar /app/web.jar server /app/config/config.yaml

EXPOSE 8080 8443 8081 8444

**Comment by siruslan** [comment]
Seems like it's Jelastic related only as it runs Docker images with additional enhancements for legacy apps that depend on init, ssh, cron, logrotate, ftp, fail2ban and others.

+@ihorman +@xzkutor +@mrAlexZT for clarifying the issue

**Comment by derkd** [comment]
Ok is there something I can do about this? Should I use a different base image i.e.?

**Comment by mrAlexZT** [comment]
@derkd we have tried to reproduce mentioned issue on the [Mirhosting](https://jelastic.cloud/?query=hoster_mirhosting) platform and unable to reproduce it. Could you please provide us with more details about docker image and hosting provider where you face the issue with the deploy?

**Comment by derkd** [comment]
@mrAlexZT 

please try the docker image below. In this image I have not added:

```
#comment all tty stuff in /etc/inittab
RUN sed -i -- 's/tty/#tty/g' /etc/inittab
```

If you check the /var/log/messages you will see a constant stream of getty respawn messages.


```
FROM openjdk:8u121-jre-alpine

ENV JELASTIC_EXPOSE FALSE
ENV TIMEZONE Europe/Amsterdam
RUN apk add --update tzdata && \
	cp /usr/share/zoneinfo/${TIMEZONE} /etc/localtime && \
	echo "${TIMEZONE}" > /etc/timezone && \
	apk del tzdata

COPY web.jar /app/web.jar

RUN mkdir /app/config
VOLUME ["/app/config"]

RUN mkdir /app/invoices
VOLUME ["/app/invoices"]

RUN mkdir /app/credits
VOLUME ["/app/credits"]

RUN mkdir /var/log/app
VOLUME ["/var/log/app"]

CMD java -jar /app/web.jar server /app/config/configuration.yaml

EXPOSE 8080 8443
```

**Comment by mrAlexZT** [comment]
This is a Virtuozzo bug **#PSBM-64255** which will be fixed in one of future Virtuozo's releases.
We have applied a hotfix from Jelastic's side on all platforms and now docker redeploy works without any problems.

**Comment by yosifkit** [comment]
Closing since it looks like this was fixed in Jelastic Cloud.
